### ISSUE TITLE
seq2seq.AttentionWrapper cannot implement Bahdanau model (RNNsearch)

### ISSUE THREAD

**Issue by jagthebeetle** [tensorflow/tensorflow#11365]
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes, but there is no stock example script
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: OSX 10.12.5
- **TensorFlow installed from (source or binary)**: source (via virtualenv/pip3)
- **TensorFlow version (use command below)**:v1.2.0-5-g435cdfc 1.2.1
- **Python version**: 3.6.1
- **Bazel version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**:  N/A

### Describe the problem
[RNNsearch](https://arxiv.org/pdf/1409.0473.pdf) has two different RNN cells in it (the encoder and decoder). At a given time step, the inputs to the attention mechanism are all the encoder's outputs (the "annotations" in the Bahdanau paper, aka the `memory` of the `BahdanauAttention` constructor) and the decoder's previous state.

Crucially, `AttentionWrapper.call` is running its input through the passed in `cell` before applying `AttentionMechanism` to the cell's output. ([Docstring](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/AttentionWrapper#call); source confirms this.)

This precludes the encoder cell from being passed to AttentionWrapper, because AttentionWrapper depends on BahdanauAttention, whose `memory` depends on the encoder cell's output. (So you would need to run the encoder once to get the `memory` and `AttentionWrapper` would run it again.)

But the decoder cell can also not be the cell passed in, because in the Bahdanau paper, the decoder's output is not used by the attention mechanism at all. Moreover, neither cell takes the previous step's attention (in Bahdanau).

I'm hoping I've misunderstood, but currently the API doesn't seem like it can be made to align with the paper, which would be sort of curious – but perhaps intentional!

### Source code / logs
    bahdanau = tf.contrib.seq2seq.BahdanauAttention(
        num_units=params['ATTENTION_SIZE'],
        memory=annotations,  # annotations, _ = tf.nn.static_rnn(encoder, time_major_input)
        normalize=False,
        name='BahdanauAttention')
    decoder = tf.nn.rnn_cell.BasicLSTMCell(
        params['DECODER_SIZE'],
        forget_bias=1.0)
    attn_cell = tf.contrib.seq2seq.AttentionWrapper(
        cell=decoder,
        attention_mechanism=bahdanau,
        output_attention=False,
        name="AttentionWrappedDecoder")

**Comment by jart** [comment]
@ebrevdo I'm noticing you've worked on seq2seq recently. Here's a well-communicated issue for you.

**Comment by ebrevdo** [comment]
You're meant to wrap the decoder cell in the AttentionWrapper.  The previous attention is passed via the *state* of the AttentionWrapper if you use the argument `output_attention` ([here](https://github.com/tensorflow/tensorflow/blob/afa03e2fd91a2d4a7dc960c5ec4ee448c808d026/tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py#L994)).  In practice we haven't seen too much of a difference when training if using one or the other.

**Comment by jagthebeetle** [comment]
My training results have not shown the difference to be that big either, but the edge cases worried me.

The difference at the first step (step1) of decoding is between:
```
state1 = cell((context1=attn(state0)), state0) # RNNSearch
context1 = attn((state1 = cell(context0, state0))) # AttentionWrapper
```
(assuming, as in GRU cells, that states and outputs are the same)

On further inspection, the two turn out to be the same if I provide the initial state for AttentionWrapper as:
```
AttentionWrapperState(
    attention=attn(_cell.zero_state()) # score with bahdanau and calculate context from alignments
    cell_state=_cell.zero_state())
```
Leaving the note for posterity. Thanks for the clarification!
